<a href="https://colab.research.google.com/github/azakatov/ut-ml-project-2024/blob/try-ensemble/main.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<svg width="416.16" height="24.064" viewBox="0 0 416.16 24.064" xmlns="http://www.w3.org/2000/svg"><g id="svgGroup" stroke-linecap="round" fill-rule="evenodd" font-size="9pt" stroke="#ff0000" stroke-width="0.25mm" fill="#ff0000" style="stroke:#ff0000;stroke-width:0.25mm;fill:#ff0000"><path d="M 2.88 23.68 L 0 23.68 L 0 1.28 L 4.256 1.28 L 10.464 15.456 L 16.576 1.28 L 20.928 1.28 L 20.928 23.68 L 17.888 23.68 L 17.888 5.152 L 11.904 18.624 L 8.896 18.624 L 2.88 5.152 L 2.88 23.68 Z M 249.6 23.68 L 246.72 23.68 L 246.72 0 L 249.6 0 L 249.6 9.696 Q 250.496 8.384 251.872 7.52 A 5.623 5.623 0 0 1 254.152 6.717 A 7.23 7.23 0 0 1 255.104 6.656 A 7.432 7.432 0 0 1 256.725 6.822 Q 258.355 7.186 259.312 8.352 A 5.889 5.889 0 0 1 260.465 10.713 Q 260.704 11.679 260.704 12.832 L 260.704 23.68 L 257.824 23.68 L 257.824 13.152 A 6.28 6.28 0 0 0 257.717 11.955 Q 257.525 10.969 256.992 10.272 Q 256.16 9.184 254.656 9.184 A 4.233 4.233 0 0 0 252.508 9.781 A 5.627 5.627 0 0 0 251.856 10.224 Q 250.528 11.264 249.6 12.672 L 249.6 23.68 Z M 59.584 22.08 L 57.6 24 L 48.992 15.136 L 48.992 23.68 L 46.112 23.68 L 46.112 0 L 48.992 0 L 48.992 13.536 L 56.512 6.72 L 58.176 8.544 L 51.744 14.24 L 59.584 22.08 Z M 103.808 17.856 L 103.808 7.04 L 106.688 7.04 L 106.688 17.472 A 7.354 7.354 0 0 0 106.769 18.603 Q 106.861 19.194 107.058 19.67 A 3.244 3.244 0 0 0 107.52 20.464 A 2.689 2.689 0 0 0 109.156 21.431 A 4.098 4.098 0 0 0 109.952 21.504 Q 111.264 21.504 112.496 20.976 Q 113.728 20.448 114.688 19.552 L 114.688 7.04 L 117.568 7.04 L 117.568 19.36 A 4.992 4.992 0 0 0 117.601 19.953 Q 117.707 20.837 118.16 21.168 Q 118.752 21.6 119.456 21.6 L 118.848 23.84 Q 115.872 23.84 115.168 21.472 Q 114.048 22.624 112.576 23.344 Q 111.104 24.064 109.408 24.064 Q 107.968 24.064 106.688 23.456 A 4.621 4.621 0 0 1 104.925 21.972 A 5.845 5.845 0 0 1 104.608 21.488 A 5.197 5.197 0 0 1 104.075 20.178 Q 103.908 19.548 103.846 18.798 A 11.319 11.319 0 0 1 103.808 17.856 Z M 215.552 23.68 L 212.672 23.68 L 212.672 7.04 L 215.456 7.04 L 215.456 9.824 Q 216.352 8.48 217.744 7.568 Q 219.136 6.656 221.056 6.656 A 7.432 7.432 0 0 1 222.677 6.822 Q 224.307 7.186 225.264 8.352 A 5.889 5.889 0 0 1 226.417 10.713 Q 226.656 11.679 226.656 12.832 L 226.656 23.68 L 223.776 23.68 L 223.776 13.152 A 6.165 6.165 0 0 0 223.666 11.955 Q 223.471 10.969 222.928 10.272 A 2.745 2.745 0 0 0 220.919 9.199 A 3.86 3.86 0 0 0 220.576 9.184 Q 219.104 9.184 217.792 10.224 Q 216.48 11.264 215.552 12.672 L 215.552 23.68 Z M 87.552 22.336 L 88.768 19.936 A 5.938 5.938 0 0 0 89.774 20.616 Q 90.268 20.884 90.855 21.101 A 9.843 9.843 0 0 0 90.864 21.104 A 7.703 7.703 0 0 0 93.3 21.564 A 8.789 8.789 0 0 0 93.568 21.568 Q 95.294 21.568 96.203 21.019 A 2.476 2.476 0 0 0 96.272 20.976 A 2.384 2.384 0 0 0 96.763 20.562 A 1.617 1.617 0 0 0 97.184 19.456 A 2.279 2.279 0 0 0 97.084 18.769 A 1.988 1.988 0 0 0 96.848 18.272 A 2.043 2.043 0 0 0 96.551 17.922 Q 96.204 17.59 95.616 17.264 A 9.108 9.108 0 0 0 94.978 16.946 Q 94.301 16.638 93.344 16.303 A 33.156 33.156 0 0 0 93.024 16.192 A 14.917 14.917 0 0 1 91.617 15.634 Q 90.305 15.03 89.52 14.288 A 3.624 3.624 0 0 1 88.466 12.369 A 5.374 5.374 0 0 1 88.352 11.232 Q 88.352 9.312 90.032 7.984 A 5.685 5.685 0 0 1 91.964 7.009 Q 92.782 6.766 93.763 6.69 A 12.009 12.009 0 0 1 94.688 6.656 A 13.054 13.054 0 0 1 96.111 6.73 Q 96.873 6.814 97.536 6.992 A 12.406 12.406 0 0 1 98.859 7.426 A 9.955 9.955 0 0 1 99.776 7.84 L 98.976 10.208 A 7.849 7.849 0 0 0 97.802 9.643 A 9.661 9.661 0 0 0 97.008 9.376 A 8.008 8.008 0 0 0 95.573 9.105 A 10.188 10.188 0 0 0 94.56 9.056 Q 93.493 9.056 92.754 9.333 A 2.723 2.723 0 0 0 92.032 9.728 A 2.983 2.983 0 0 0 91.598 10.136 Q 91.168 10.636 91.168 11.2 Q 91.168 11.968 91.84 12.512 Q 92.512 13.056 94.464 13.696 A 21.681 21.681 0 0 1 96.084 14.303 Q 97.624 14.958 98.507 15.689 A 4.783 4.783 0 0 1 98.88 16.032 A 4.351 4.351 0 0 1 100.123 18.991 A 5.727 5.727 0 0 1 100.128 19.232 Q 100.128 21.472 98.32 22.768 A 6.417 6.417 0 0 1 96.279 23.719 Q 95.427 23.956 94.42 24.03 A 12.935 12.935 0 0 1 93.472 24.064 A 12.797 12.797 0 0 1 91.596 23.932 A 9.903 9.903 0 0 1 90.064 23.584 A 9.883 9.883 0 0 1 88.835 23.109 Q 88.11 22.768 87.552 22.336 Z M 281.824 22.336 L 283.04 19.936 A 5.938 5.938 0 0 0 284.046 20.616 Q 284.54 20.884 285.127 21.101 A 9.843 9.843 0 0 0 285.136 21.104 A 7.703 7.703 0 0 0 287.572 21.564 A 8.789 8.789 0 0 0 287.84 21.568 Q 289.566 21.568 290.475 21.019 A 2.476 2.476 0 0 0 290.544 20.976 A 2.384 2.384 0 0 0 291.035 20.562 A 1.617 1.617 0 0 0 291.456 19.456 A 2.279 2.279 0 0 0 291.356 18.769 A 1.988 1.988 0 0 0 291.12 18.272 A 2.043 2.043 0 0 0 290.823 17.922 Q 290.476 17.59 289.888 17.264 A 9.108 9.108 0 0 0 289.25 16.946 Q 288.573 16.638 287.616 16.303 A 33.156 33.156 0 0 0 287.296 16.192 A 14.917 14.917 0 0 1 285.889 15.634 Q 284.577 15.03 283.792 14.288 A 3.624 3.624 0 0 1 282.738 12.369 A 5.374 5.374 0 0 1 282.624 11.232 Q 282.624 9.312 284.304 7.984 A 5.685 5.685 0 0 1 286.236 7.009 Q 287.054 6.766 288.035 6.69 A 12.009 12.009 0 0 1 288.96 6.656 A 13.054 13.054 0 0 1 290.383 6.73 Q 291.145 6.814 291.808 6.992 A 12.406 12.406 0 0 1 293.131 7.426 A 9.955 9.955 0 0 1 294.048 7.84 L 293.248 10.208 A 7.849 7.849 0 0 0 292.074 9.643 A 9.661 9.661 0 0 0 291.28 9.376 A 8.008 8.008 0 0 0 289.845 9.105 A 10.188 10.188 0 0 0 288.832 9.056 Q 287.765 9.056 287.026 9.333 A 2.723 2.723 0 0 0 286.304 9.728 A 2.983 2.983 0 0 0 285.87 10.136 Q 285.44 10.636 285.44 11.2 Q 285.44 11.968 286.112 12.512 Q 286.784 13.056 288.736 13.696 A 21.681 21.681 0 0 1 290.356 14.303 Q 291.896 14.958 292.779 15.689 A 4.783 4.783 0 0 1 293.152 16.032 A 4.351 4.351 0 0 1 294.395 18.991 A 5.727 5.727 0 0 1 294.4 19.232 Q 294.4 21.472 292.592 22.768 A 6.417 6.417 0 0 1 290.551 23.719 Q 289.699 23.956 288.692 24.03 A 12.935 12.935 0 0 1 287.744 24.064 A 12.797 12.797 0 0 1 285.868 23.932 A 9.903 9.903 0 0 1 284.336 23.584 A 9.883 9.883 0 0 1 283.107 23.109 Q 282.382 22.768 281.824 22.336 Z M 164.8 22.304 L 164.8 23.68 L 162.112 23.68 L 162.112 0 L 164.992 0 L 164.992 8.768 A 4.964 4.964 0 0 1 165.596 8.189 Q 165.921 7.923 166.328 7.663 A 10.661 10.661 0 0 1 166.864 7.344 A 5.307 5.307 0 0 1 168.538 6.761 A 7.176 7.176 0 0 1 169.792 6.656 Q 171.712 6.656 173.392 7.664 A 7.145 7.145 0 0 1 175.689 9.893 A 8.789 8.789 0 0 1 176.096 10.576 Q 177.12 12.48 177.12 15.2 Q 177.12 17.184 176.48 18.816 A 9.397 9.397 0 0 1 175.468 20.737 A 8.154 8.154 0 0 1 174.768 21.616 Q 173.696 22.784 172.352 23.424 A 6.402 6.402 0 0 1 169.995 24.051 A 6.147 6.147 0 0 1 169.6 24.064 Q 168.16 24.064 166.896 23.536 A 9.688 9.688 0 0 1 165.839 23.021 Q 165.31 22.72 164.886 22.376 A 5.836 5.836 0 0 1 164.8 22.304 Z M 75.456 16.544 L 63.808 16.544 Q 64.076 18.72 65.261 20.008 A 4.794 4.794 0 0 0 65.44 20.192 A 4.526 4.526 0 0 0 67.88 21.411 A 6.407 6.407 0 0 0 68.992 21.504 Q 70.432 21.504 71.584 21.2 Q 72.736 20.896 73.792 20.416 L 74.464 22.88 A 13.046 13.046 0 0 1 73.077 23.4 A 16.596 16.596 0 0 1 71.872 23.728 A 12.453 12.453 0 0 1 70.243 23.99 A 16.648 16.648 0 0 1 68.64 24.064 A 9.062 9.062 0 0 1 66.161 23.742 A 6.755 6.755 0 0 1 62.96 21.776 A 7.527 7.527 0 0 1 61.294 18.723 Q 60.972 17.604 60.891 16.272 A 15.08 15.08 0 0 1 60.864 15.36 Q 60.864 12.864 61.808 10.896 A 7.885 7.885 0 0 1 63.413 8.63 A 7.324 7.324 0 0 1 64.448 7.792 A 6.735 6.735 0 0 1 67.731 6.681 A 8.386 8.386 0 0 1 68.384 6.656 Q 70.784 6.656 72.368 7.696 Q 73.952 8.736 74.752 10.496 Q 75.552 12.256 75.552 14.368 A 24.065 24.065 0 0 1 75.459 16.511 A 22.061 22.061 0 0 1 75.456 16.544 Z M 148.576 16.544 L 136.928 16.544 Q 137.196 18.72 138.381 20.008 A 4.794 4.794 0 0 0 138.56 20.192 A 4.526 4.526 0 0 0 141 21.411 A 6.407 6.407 0 0 0 142.112 21.504 Q 143.552 21.504 144.704 21.2 Q 145.856 20.896 146.912 20.416 L 147.584 22.88 A 13.046 13.046 0 0 1 146.197 23.4 A 16.596 16.596 0 0 1 144.992 23.728 A 12.453 12.453 0 0 1 143.363 23.99 A 16.648 16.648 0 0 1 141.76 24.064 A 9.062 9.062 0 0 1 139.281 23.742 A 6.755 6.755 0 0 1 136.08 21.776 A 7.527 7.527 0 0 1 134.414 18.723 Q 134.092 17.604 134.011 16.272 A 15.08 15.08 0 0 1 133.984 15.36 Q 133.984 12.864 134.928 10.896 A 7.885 7.885 0 0 1 136.533 8.63 A 7.324 7.324 0 0 1 137.568 7.792 A 6.735 6.735 0 0 1 140.851 6.681 A 8.386 8.386 0 0 1 141.504 6.656 Q 143.904 6.656 145.488 7.696 Q 147.072 8.736 147.872 10.496 Q 148.672 12.256 148.672 14.368 A 24.065 24.065 0 0 1 148.579 16.511 A 22.061 22.061 0 0 1 148.576 16.544 Z M 379.808 16.544 L 368.16 16.544 Q 368.428 18.72 369.613 20.008 A 4.794 4.794 0 0 0 369.792 20.192 A 4.526 4.526 0 0 0 372.232 21.411 A 6.407 6.407 0 0 0 373.344 21.504 Q 374.784 21.504 375.936 21.2 Q 377.088 20.896 378.144 20.416 L 378.816 22.88 A 13.046 13.046 0 0 1 377.429 23.4 A 16.596 16.596 0 0 1 376.224 23.728 A 12.453 12.453 0 0 1 374.595 23.99 A 16.648 16.648 0 0 1 372.992 24.064 A 9.062 9.062 0 0 1 370.513 23.742 A 6.755 6.755 0 0 1 367.312 21.776 A 7.527 7.527 0 0 1 365.646 18.723 Q 365.324 17.604 365.243 16.272 A 15.08 15.08 0 0 1 365.216 15.36 Q 365.216 12.864 366.16 10.896 A 7.885 7.885 0 0 1 367.765 8.63 A 7.324 7.324 0 0 1 368.8 7.792 A 6.735 6.735 0 0 1 372.083 6.681 A 8.386 8.386 0 0 1 372.736 6.656 Q 375.136 6.656 376.72 7.696 Q 378.304 8.736 379.104 10.496 Q 379.904 12.256 379.904 14.368 A 24.065 24.065 0 0 1 379.811 16.511 A 22.061 22.061 0 0 1 379.808 16.544 Z M 399.488 17.952 L 399.488 10.144 L 396.608 10.144 L 396.608 7.648 L 399.552 7.648 L 400.16 2.4 L 402.368 2.4 L 402.368 7.648 L 406.976 7.648 L 406.976 10.144 L 402.368 10.144 L 402.368 18.176 A 6.899 6.899 0 0 0 402.422 19.071 Q 402.573 20.225 403.152 20.768 Q 403.936 21.504 405.024 21.504 Q 405.824 21.504 406.544 21.248 Q 407.264 20.992 407.872 20.64 L 408.672 22.944 Q 408.032 23.328 406.944 23.696 Q 405.856 24.064 404.672 24.064 A 5.865 5.865 0 0 1 403.014 23.841 A 4.459 4.459 0 0 1 400.88 22.48 A 5.157 5.157 0 0 1 399.847 20.591 Q 399.603 19.814 399.525 18.874 A 11.129 11.129 0 0 1 399.488 17.952 Z M 243.104 7.648 L 242.24 10.08 Q 241.344 9.632 240.528 9.424 A 6.574 6.574 0 0 0 239.554 9.256 A 8.434 8.434 0 0 0 238.72 9.216 A 5.085 5.085 0 0 0 236.399 9.742 A 5.162 5.162 0 0 0 234.96 10.832 A 5.355 5.355 0 0 0 233.691 13.185 Q 233.471 14.02 233.444 15.01 A 10.328 10.328 0 0 0 233.44 15.296 Q 233.44 17.024 234.08 18.432 A 5.94 5.94 0 0 0 235.011 19.897 A 5.208 5.208 0 0 0 235.856 20.672 Q 236.992 21.504 238.528 21.504 Q 239.744 21.504 240.608 21.232 Q 241.469 20.961 242.456 20.452 A 16.647 16.647 0 0 0 242.464 20.448 L 243.36 22.88 A 9.132 9.132 0 0 1 241.883 23.522 A 10.973 10.973 0 0 1 241.136 23.744 Q 239.923 24.059 238.385 24.064 A 15.314 15.314 0 0 1 238.336 24.064 A 8.577 8.577 0 0 1 236.084 23.779 A 7.064 7.064 0 0 1 234.192 22.96 Q 232.416 21.856 231.456 19.888 A 9.396 9.396 0 0 1 230.612 17.068 A 12.167 12.167 0 0 1 230.496 15.36 A 10.252 10.252 0 0 1 230.916 12.389 A 9.215 9.215 0 0 1 231.472 10.976 Q 232.448 8.992 234.272 7.824 A 7.382 7.382 0 0 1 237.366 6.729 A 9.475 9.475 0 0 1 238.56 6.656 Q 239.844 6.656 240.849 6.869 A 7.67 7.67 0 0 1 241.04 6.912 Q 242.112 7.168 243.104 7.648 Z M 319.328 7.648 L 318.464 10.08 Q 317.568 9.632 316.752 9.424 A 6.574 6.574 0 0 0 315.778 9.256 A 8.434 8.434 0 0 0 314.944 9.216 A 5.085 5.085 0 0 0 312.623 9.742 A 5.162 5.162 0 0 0 311.184 10.832 A 5.355 5.355 0 0 0 309.915 13.185 Q 309.695 14.02 309.668 15.01 A 10.328 10.328 0 0 0 309.664 15.296 Q 309.664 17.024 310.304 18.432 A 5.94 5.94 0 0 0 311.235 19.897 A 5.208 5.208 0 0 0 312.08 20.672 Q 313.216 21.504 314.752 21.504 Q 315.968 21.504 316.832 21.232 Q 317.693 20.961 318.68 20.452 A 16.647 16.647 0 0 0 318.688 20.448 L 319.584 22.88 A 9.132 9.132 0 0 1 318.107 23.522 A 10.973 10.973 0 0 1 317.36 23.744 Q 316.147 24.059 314.609 24.064 A 15.314 15.314 0 0 1 314.56 24.064 A 8.577 8.577 0 0 1 312.308 23.779 A 7.064 7.064 0 0 1 310.416 22.96 Q 308.64 21.856 307.68 19.888 A 9.396 9.396 0 0 1 306.836 17.068 A 12.167 12.167 0 0 1 306.72 15.36 A 10.252 10.252 0 0 1 307.14 12.389 A 9.215 9.215 0 0 1 307.696 10.976 Q 308.672 8.992 310.496 7.824 A 7.382 7.382 0 0 1 313.59 6.729 A 9.475 9.475 0 0 1 314.784 6.656 Q 316.068 6.656 317.073 6.869 A 7.67 7.67 0 0 1 317.264 6.912 Q 318.336 7.168 319.328 7.648 Z M 395.232 7.648 L 394.368 10.08 Q 393.472 9.632 392.656 9.424 A 6.574 6.574 0 0 0 391.682 9.256 A 8.434 8.434 0 0 0 390.848 9.216 A 5.085 5.085 0 0 0 388.527 9.742 A 5.162 5.162 0 0 0 387.088 10.832 A 5.355 5.355 0 0 0 385.819 13.185 Q 385.599 14.02 385.572 15.01 A 10.328 10.328 0 0 0 385.568 15.296 Q 385.568 17.024 386.208 18.432 A 5.94 5.94 0 0 0 387.139 19.897 A 5.208 5.208 0 0 0 387.984 20.672 Q 389.12 21.504 390.656 21.504 Q 391.872 21.504 392.736 21.232 Q 393.597 20.961 394.584 20.452 A 16.647 16.647 0 0 0 394.592 20.448 L 395.488 22.88 A 9.132 9.132 0 0 1 394.011 23.522 A 10.973 10.973 0 0 1 393.264 23.744 Q 392.051 24.059 390.513 24.064 A 15.314 15.314 0 0 1 390.464 24.064 A 8.577 8.577 0 0 1 388.212 23.779 A 7.064 7.064 0 0 1 386.32 22.96 Q 384.544 21.856 383.584 19.888 A 9.396 9.396 0 0 1 382.74 17.068 A 12.167 12.167 0 0 1 382.624 15.36 A 10.252 10.252 0 0 1 383.044 12.389 A 9.215 9.215 0 0 1 383.6 10.976 Q 384.576 8.992 386.4 7.824 A 7.382 7.382 0 0 1 389.494 6.729 A 9.475 9.475 0 0 1 390.688 6.656 Q 391.972 6.656 392.977 6.869 A 7.67 7.67 0 0 1 393.168 6.912 Q 394.24 7.168 395.232 7.648 Z M 37.76 8.384 L 37.76 7.04 L 40.384 7.04 L 40.384 19.648 A 3.888 3.888 0 0 0 40.417 20.171 Q 40.499 20.774 40.788 21.079 A 0.949 0.949 0 0 0 40.976 21.232 Q 41.568 21.6 42.272 21.6 L 41.664 23.84 A 5.671 5.671 0 0 1 40.299 23.689 Q 38.496 23.241 37.985 21.466 A 4.349 4.349 0 0 1 37.952 21.344 A 6.818 6.818 0 0 1 36.957 22.44 A 9.119 9.119 0 0 1 35.952 23.216 Q 34.764 24.013 32.969 24.061 A 8.706 8.706 0 0 1 32.736 24.064 Q 30.656 24.064 28.96 23.008 A 7.343 7.343 0 0 1 26.588 20.6 A 8.935 8.935 0 0 1 26.256 20.016 A 8.713 8.713 0 0 1 25.414 17.474 A 11.726 11.726 0 0 1 25.248 15.456 A 10.605 10.605 0 0 1 25.567 12.81 A 8.95 8.95 0 0 1 26.256 10.976 Q 27.264 8.992 29.04 7.824 Q 30.816 6.656 33.12 6.656 A 7.227 7.227 0 0 1 34.654 6.813 A 5.985 5.985 0 0 1 35.728 7.152 A 8.321 8.321 0 0 1 37.1 7.886 A 7.073 7.073 0 0 1 37.76 8.384 Z M 204.32 8.384 L 204.32 7.04 L 206.944 7.04 L 206.944 19.648 A 3.888 3.888 0 0 0 206.977 20.171 Q 207.059 20.774 207.348 21.079 A 0.949 0.949 0 0 0 207.536 21.232 Q 208.128 21.6 208.832 21.6 L 208.224 23.84 A 5.671 5.671 0 0 1 206.859 23.689 Q 205.056 23.241 204.545 21.466 A 4.349 4.349 0 0 1 204.512 21.344 A 6.818 6.818 0 0 1 203.517 22.44 A 9.119 9.119 0 0 1 202.512 23.216 Q 201.324 24.013 199.529 24.061 A 8.706 8.706 0 0 1 199.296 24.064 Q 197.216 24.064 195.52 23.008 A 7.343 7.343 0 0 1 193.148 20.6 A 8.935 8.935 0 0 1 192.816 20.016 A 8.713 8.713 0 0 1 191.974 17.474 A 11.726 11.726 0 0 1 191.808 15.456 A 10.605 10.605 0 0 1 192.127 12.81 A 8.95 8.95 0 0 1 192.816 10.976 Q 193.824 8.992 195.6 7.824 Q 197.376 6.656 199.68 6.656 A 7.227 7.227 0 0 1 201.214 6.813 A 5.985 5.985 0 0 1 202.288 7.152 A 8.321 8.321 0 0 1 203.66 7.886 A 7.073 7.073 0 0 1 204.32 8.384 Z M 126.176 23.68 L 123.296 23.68 L 123.296 7.04 L 126.08 7.04 L 126.08 10.656 Q 126.496 9.6 127.2 8.688 Q 127.904 7.776 128.96 7.216 Q 130.016 6.656 131.424 6.656 Q 131.904 6.656 132.384 6.704 Q 132.864 6.752 133.184 6.848 L 132.32 9.824 A 3.124 3.124 0 0 0 131.633 9.628 Q 131.316 9.574 130.955 9.569 A 5.309 5.309 0 0 0 130.88 9.568 Q 129.696 9.568 128.624 10.24 Q 127.552 10.912 126.864 12.32 A 6.183 6.183 0 0 0 126.394 13.699 Q 126.246 14.366 126.198 15.145 A 12.509 12.509 0 0 0 126.176 15.904 L 126.176 23.68 Z M 184 23.68 L 181.12 23.68 L 181.12 7.04 L 183.904 7.04 L 183.904 10.656 Q 184.32 9.6 185.024 8.688 Q 185.728 7.776 186.784 7.216 Q 187.84 6.656 189.248 6.656 Q 189.728 6.656 190.208 6.704 Q 190.688 6.752 191.008 6.848 L 190.144 9.824 A 3.124 3.124 0 0 0 189.457 9.628 Q 189.14 9.574 188.779 9.569 A 5.309 5.309 0 0 0 188.704 9.568 Q 187.52 9.568 186.448 10.24 Q 185.376 10.912 184.688 12.32 A 6.183 6.183 0 0 0 184.218 13.699 Q 184.07 14.366 184.022 15.145 A 12.509 12.509 0 0 0 184 15.904 L 184 23.68 Z M 344.64 23.68 L 341.76 23.68 L 341.76 7.04 L 344.544 7.04 L 344.544 10.656 Q 344.96 9.6 345.664 8.688 Q 346.368 7.776 347.424 7.216 Q 348.48 6.656 349.888 6.656 Q 350.368 6.656 350.848 6.704 Q 351.328 6.752 351.648 6.848 L 350.784 9.824 A 3.124 3.124 0 0 0 350.097 9.628 Q 349.78 9.574 349.419 9.569 A 5.309 5.309 0 0 0 349.344 9.568 Q 348.16 9.568 347.088 10.24 Q 346.016 10.912 345.328 12.32 A 6.183 6.183 0 0 0 344.858 13.699 Q 344.71 14.366 344.662 15.145 A 12.509 12.509 0 0 0 344.64 15.904 L 344.64 23.68 Z M 357.408 23.68 L 354.528 23.68 L 354.528 7.04 L 357.312 7.04 L 357.312 10.656 Q 357.728 9.6 358.432 8.688 Q 359.136 7.776 360.192 7.216 Q 361.248 6.656 362.656 6.656 Q 363.136 6.656 363.616 6.704 Q 364.096 6.752 364.416 6.848 L 363.552 9.824 A 3.124 3.124 0 0 0 362.865 9.628 Q 362.548 9.574 362.187 9.569 A 5.309 5.309 0 0 0 362.112 9.568 Q 360.928 9.568 359.856 10.24 Q 358.784 10.912 358.096 12.32 A 6.183 6.183 0 0 0 357.626 13.699 Q 357.478 14.366 357.43 15.145 A 12.509 12.509 0 0 0 357.408 15.904 L 357.408 23.68 Z M 326.868 23.635 A 8.709 8.709 0 0 0 329.632 24.064 A 9.832 9.832 0 0 0 330.36 24.037 A 8.082 8.082 0 0 0 333.872 22.976 Q 335.744 21.888 336.832 19.92 A 8.402 8.402 0 0 0 337.681 17.635 A 10.939 10.939 0 0 0 337.92 15.296 A 11.053 11.053 0 0 0 337.765 13.405 A 8.254 8.254 0 0 0 336.816 10.704 Q 335.712 8.768 333.84 7.712 A 7.952 7.952 0 0 0 332.355 7.063 A 8.876 8.876 0 0 0 329.632 6.656 Q 327.296 6.656 325.408 7.712 Q 323.52 8.768 322.432 10.72 A 8.327 8.327 0 0 0 321.552 13.149 A 11.202 11.202 0 0 0 321.344 15.36 A 11.142 11.142 0 0 0 321.429 16.753 A 8.784 8.784 0 0 0 322.416 19.888 Q 323.488 21.856 325.376 22.96 A 7.859 7.859 0 0 0 326.868 23.635 Z M 278.144 23.68 L 275.264 23.68 L 275.264 7.04 L 278.144 7.04 L 278.144 23.68 Z M 415.168 17.056 L 412.8 17.056 L 412.384 8.32 L 412.384 0.352 L 415.584 0.352 L 415.584 8.32 L 415.168 17.056 Z M 329.632 21.504 A 5.997 5.997 0 0 0 331.304 21.283 A 4.48 4.48 0 0 0 333.568 19.792 Q 334.976 18.08 334.976 15.36 A 6.84 6.84 0 0 0 334.584 13.028 A 6.404 6.404 0 0 0 334.272 12.304 Q 333.568 10.912 332.368 10.064 Q 331.168 9.216 329.632 9.216 A 6.16 6.16 0 0 0 327.96 9.431 A 4.492 4.492 0 0 0 325.696 10.88 A 5.721 5.721 0 0 0 324.52 13.24 Q 324.31 14.09 324.29 15.088 A 10.518 10.518 0 0 0 324.288 15.296 Q 324.288 16.96 324.992 18.384 A 6.22 6.22 0 0 0 326.122 19.997 A 5.641 5.641 0 0 0 326.896 20.656 Q 328.096 21.504 329.632 21.504 Z M 37.504 19.424 L 37.504 10.496 Q 36.64 9.888 35.6 9.52 Q 34.56 9.152 33.408 9.152 A 5.096 5.096 0 0 0 31.817 9.393 A 4.488 4.488 0 0 0 30.704 9.936 A 5.216 5.216 0 0 0 29.098 11.672 A 6.388 6.388 0 0 0 28.864 12.112 A 6.612 6.612 0 0 0 28.303 13.929 A 8.857 8.857 0 0 0 28.192 15.36 Q 28.192 17.152 28.864 18.544 A 5.633 5.633 0 0 0 29.858 19.999 A 5.023 5.023 0 0 0 30.72 20.72 Q 31.904 21.504 33.376 21.504 Q 34.592 21.504 35.68 20.912 Q 36.768 20.32 37.504 19.424 Z M 204.064 19.424 L 204.064 10.496 Q 203.2 9.888 202.16 9.52 Q 201.12 9.152 199.968 9.152 A 5.096 5.096 0 0 0 198.377 9.393 A 4.488 4.488 0 0 0 197.264 9.936 A 5.216 5.216 0 0 0 195.658 11.672 A 6.388 6.388 0 0 0 195.424 12.112 A 6.612 6.612 0 0 0 194.863 13.929 A 8.857 8.857 0 0 0 194.752 15.36 Q 194.752 17.152 195.424 18.544 A 5.633 5.633 0 0 0 196.418 19.999 A 5.023 5.023 0 0 0 197.28 20.72 Q 198.464 21.504 199.936 21.504 Q 201.152 21.504 202.24 20.912 Q 203.328 20.32 204.064 19.424 Z M 164.992 11.328 L 164.992 20.224 Q 165.728 20.8 166.72 21.152 Q 167.712 21.504 168.832 21.504 A 5.437 5.437 0 0 0 170.437 21.274 A 4.719 4.719 0 0 0 171.6 20.736 A 5.104 5.104 0 0 0 173.248 19.005 A 6.287 6.287 0 0 0 173.488 18.56 A 6.678 6.678 0 0 0 174.079 16.625 A 8.793 8.793 0 0 0 174.176 15.296 Q 174.176 13.444 173.536 12.099 A 5.547 5.547 0 0 0 173.488 12 Q 172.8 10.624 171.664 9.888 Q 170.528 9.152 169.152 9.152 A 4.717 4.717 0 0 0 167.351 9.492 A 4.497 4.497 0 0 0 166.672 9.84 A 7.537 7.537 0 0 0 165.862 10.417 Q 165.466 10.744 165.175 11.093 A 4.141 4.141 0 0 0 164.992 11.328 Z M 63.776 14.208 L 72.896 14.208 Q 72.896 12.442 72.255 11.266 A 3.975 3.975 0 0 0 71.68 10.464 A 3.996 3.996 0 0 0 69.453 9.256 A 5.886 5.886 0 0 0 68.32 9.152 A 4.346 4.346 0 0 0 66.684 9.451 A 4.008 4.008 0 0 0 65.248 10.448 A 4.853 4.853 0 0 0 64.307 11.949 Q 64.037 12.608 63.889 13.415 A 10.247 10.247 0 0 0 63.776 14.208 Z M 136.896 14.208 L 146.016 14.208 Q 146.016 12.442 145.375 11.266 A 3.975 3.975 0 0 0 144.8 10.464 A 3.996 3.996 0 0 0 142.573 9.256 A 5.886 5.886 0 0 0 141.44 9.152 A 4.346 4.346 0 0 0 139.804 9.451 A 4.008 4.008 0 0 0 138.368 10.448 A 4.853 4.853 0 0 0 137.427 11.949 Q 137.157 12.608 137.009 13.415 A 10.247 10.247 0 0 0 136.896 14.208 Z M 368.128 14.208 L 377.248 14.208 Q 377.248 12.442 376.607 11.266 A 3.975 3.975 0 0 0 376.032 10.464 A 3.996 3.996 0 0 0 373.805 9.256 A 5.886 5.886 0 0 0 372.672 9.152 A 4.346 4.346 0 0 0 371.036 9.451 A 4.008 4.008 0 0 0 369.6 10.448 A 4.853 4.853 0 0 0 368.659 11.949 Q 368.389 12.608 368.241 13.415 A 10.247 10.247 0 0 0 368.128 14.208 Z M 412.532 23.444 A 2.054 2.054 0 0 0 414.016 24.064 A 2.754 2.754 0 0 0 414.215 24.057 A 2.067 2.067 0 0 0 415.552 23.472 A 1.93 1.93 0 0 0 416.062 22.633 A 2.462 2.462 0 0 0 416.16 21.92 A 1.992 1.992 0 0 0 415.975 21.065 A 2.226 2.226 0 0 0 415.504 20.416 A 2.713 2.713 0 0 0 415.337 20.266 A 2.024 2.024 0 0 0 414.016 19.776 A 2.82 2.82 0 0 0 413.423 19.835 A 1.913 1.913 0 0 0 412.448 20.352 A 1.897 1.897 0 0 0 411.946 21.261 A 2.76 2.76 0 0 0 411.872 21.92 Q 411.872 22.784 412.512 23.424 A 2.658 2.658 0 0 0 412.532 23.444 Z M 275.308 3.612 A 1.911 1.911 0 0 0 276.704 4.192 A 2.616 2.616 0 0 0 277.183 4.15 A 1.807 1.807 0 0 0 278.176 3.632 Q 278.72 3.072 278.72 2.176 A 2.325 2.325 0 0 0 278.72 2.15 A 1.939 1.939 0 0 0 278.128 0.752 A 2.366 2.366 0 0 0 278.109 0.734 A 1.939 1.939 0 0 0 276.704 0.16 A 2.644 2.644 0 0 0 276.174 0.211 A 1.812 1.812 0 0 0 275.232 0.704 A 1.788 1.788 0 0 0 274.761 1.546 A 2.556 2.556 0 0 0 274.688 2.176 A 1.917 1.917 0 0 0 274.845 2.95 A 2.126 2.126 0 0 0 275.28 3.584 A 2.391 2.391 0 0 0 275.308 3.612 Z" vector-effect="non-scaling-stroke"/></g></svg>

<a href="https://colab.research.google.com/github/azakatov/ut-ml-project-2024/blob/main/main.ipynb" target="_parent"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="117" height="20"><linearGradient id="b" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><clipPath id="a"><rect width="117" height="20" rx="3" fill="#fff"/></clipPath><g clip-path="url(#a)"><path fill="#555" d="M0 0h30v20H0z"/><path fill="#007ec6" d="M30 0h87v20H30z"/><path fill="url(#b)" d="M0 0h117v20H0z"/></g><g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110"><svg x="4px" y="0px" width="22px" height="20px" viewBox="-2 0 28 24" style="background-color: #fff;border-radius: 1px;"><path style="fill:#e8710a;" d="M1.977,16.77c-2.667-2.277-2.605-7.079,0-9.357C2.919,8.057,3.522,9.075,4.49,9.691c-1.152,1.6-1.146,3.201-0.004,4.803C3.522,15.111,2.918,16.126,1.977,16.77z"/><path style="fill:#f9ab00;" d="M12.257,17.114c-1.767-1.633-2.485-3.658-2.118-6.02c0.451-2.91,2.139-4.893,4.946-5.678c2.565-0.718,4.964-0.217,6.878,1.819c-0.884,0.743-1.707,1.547-2.434,2.446C18.488,8.827,17.319,8.435,16,8.856c-2.404,0.767-3.046,3.241-1.494,5.644c-0.241,0.275-0.493,0.541-0.721,0.826C13.295,15.939,12.511,16.3,12.257,17.114z"/><path style="fill:#e8710a;" d="M19.529,9.682c0.727-0.899,1.55-1.703,2.434-2.446c2.703,2.783,2.701,7.031-0.005,9.764c-2.648,2.674-6.936,2.725-9.701,0.115c0.254-0.814,1.038-1.175,1.528-1.788c0.228-0.285,0.48-0.552,0.721-0.826c1.053,0.916,2.254,1.268,3.6,0.83C20.502,14.551,21.151,11.927,19.529,9.682z"/><path style="fill:#f9ab00;" d="M4.49,9.691C3.522,9.075,2.919,8.057,1.977,7.413c2.209-2.398,5.721-2.942,8.476-1.355c0.555,0.32,0.719,0.606,0.285,1.128c-0.157,0.188-0.258,0.422-0.391,0.631c-0.299,0.47-0.509,1.067-0.929,1.371C8.933,9.539,8.523,8.847,8.021,8.746C6.673,8.475,5.509,8.787,4.49,9.691z"/><path style="fill:#f9ab00;" d="M1.977,16.77c0.941-0.644,1.545-1.659,2.509-2.277c1.373,1.152,2.85,1.433,4.45,0.499c0.332-0.194,0.503-0.088,0.673,0.19c0.386,0.635,0.753,1.285,1.181,1.89c0.34,0.48,0.222,0.715-0.253,1.006C7.84,19.73,4.205,19.188,1.977,16.77z"/></svg><text x="245" y="140" transform="scale(.1)" textLength="30"> </text><text x="725" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="770">Open in Colab</text><text x="725" y="140" transform="scale(.1)" textLength="770">Colab (main)</text></g> </svg></a>

# Classification with an academic success dataset

* Project members: Andres Alumets, Anton Zakatov, Pihla Järv, Muhammad Sohaib Anwar
* Github: https://github.com/azakatov/ut-ml-project-2024
* Kaggle: https://www.kaggle.com/competitions/playground-series-s4e6/overview


Purpose of the project: contribute to the reduction of academic dropout and failure in higher education, by using machine learning techniques to identify students at risk at an early stage of their academic path, so that strategies to support them can be put into place <a href="https://archive.ics.uci.edu/dataset/697/predict+students+dropout+and+academic+success">[1]</a>.

For this project we follow a workflow proposed by Will Koehrsen <a href="https://github.com/WillKoehrsen/machine-learning-project-walkthrough/blob/master/Machine%20Learning%20Project%20Part%201.ipynb">[2]</a>.

Applying this workflow, our report has the following structure:

1. [Setup](#setup)
1. [Data cleaning and formatting](#data-cleaning-and-formatting)
1. [Exploratory data analysis](#exploratory-data-analysis)
1. [Feature engineering and selection](#feature-engineering-and-selection)
1. [Baseline model establishing](#baseline-model-establishing)
1. [Models comparison](#models-comparison)
1. [Hyperparameter tuning](#hyperparameter-tuning)
1. [Model evaluation](#model-evaluation)
1. [Results interpretation](#results-interpretation)
1. [Conclusion](#conclusion)

<a name="setup"></a>
## Setup

### Common imports

In [2]:
import numpy as np
import pandas as pd

# Show all columns in df
pd.options.display.max_columns = None

### Setting up Kaggle API

In [3]:
from google.colab import files
import json

# Upload your Kaggle token to use its API
files.upload();

!mkdir /root/.kaggle/
!mv kaggle.json /root/.kaggle/kaggle.json
!chmod 600 ~/.kaggle/kaggle.json

# Seems to work without this line but breaks data loading if present. Commenting it out
#!kaggle config set -n path -v{/content}

Saving kaggle.json to kaggle.json


In [4]:
# Load data
!kaggle competitions download -c playground-series-s4e6
!unzip playground-series-s4e6.zip

  0% 0.00/3.07M [00:00<?, ?B/s]
100% 3.07M/3.07M [00:00<00:00, 135MB/s]
Archive:  playground-series-s4e6.zip
  inflating: sample_submission.csv   
  inflating: test.csv                
  inflating: train.csv               


<a name="data-cleaning-and-formatting"></a>
## Data cleaning and formatting

We start with data loading, cleaning and formatting.

In [5]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

Using the `head()` function we can get a better understanding regarding the structure of training data.

In [ ]:
train.head()

From the output above we can see that all columns are of type `int` or `float`. Thus, the values themselves are not really interpretable. However, according to the competition's dataset description, the dataset was generated based on dataset available in <a href="https://archive.ics.uci.edu/dataset/697/predict+students+dropout+and+academic+success"/>UC Irvine Machine Learning Repository</a>. In the repository there are also explanations regarding features' values. To have a better understanding about features and their values we provide these explanations in a table below.

The table below has 7 columns:
1. `Variable Name` - a name of a variable.
2. `Role` - describes whether the variable is a feature or a target.
3. `Type` - datatype of a varialbe (integer / continuous / categorical)
4. `Demographic` - this column describes whether the column describe demographic information. Empty if it does not, contains a name of a demographic characteristic if it does.
5. `Description` - describes what values of variables actually mean or how they mapped to the actual value.
6. `Units` - unused.
7. `Missing values` - represents whether there are any missing values in a column.



<table class="table my-4 w-full">
  <thead>
    <tr>
      <th>Variable Name</th>
      <th>Role</th>
      <th>Type</th>
      <th>Demographic</th>
      <th>Description</th>
      <th>Units</th>
      <th>Missing Values</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Marital Status</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Marital Status</td>
      <td>1 – single<br />
      2 – married <br />
      3 – widower <br />
      4 – divorced <br />
      5 – facto union <br />
      6 – legally separated
      </td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Application mode</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 - 1st phase - general contingent<br />
      2 - Ordinance No. 612/93<br />
      5 - 1st phase - special contingent (Azores Island)<br />
      7 - Holders of other higher courses <br />
      10 - Ordinance No. 854-B/99 <br />
      15 - International student (bachelor) <br />
      16 - 1st phase - special contingent (Madeira Island) <br />
      17 - 2nd phase - general contingent <br />
      18 - 3rd phase - general contingent <br />
      26 - Ordinance No. 533-A/99, item b2 (Different Plan) <br />
      27 - Ordinance No. 533-A/99, item b3 (Other Institution) <br />
      39 - Over 23 years old <br />
      42 - Transfer <br />
      43 - Change of course <br />
      44 - Technological specialization diploma holders <br />
      51 - Change of institution/course <br />
      53 - Short cycle diploma holders <br />
      57 - Change of institution/course (International)
      </td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Application order</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Application order (between 0 - first choice; and 9 last choice)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Course</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>33 - Biofuel Production Technologies <br />
      171 - Animation and Multimedia Design <br />
      8014 - Social Service (evening attendance) <br />
      9003 - Agronomy <br />
      9070 - Communication Design <br />
      9085 - Veterinary Nursing <br />
      9119 - Informatics Engineering <br />
      9130 - Equinculture <br />
      9147 - Management <br />
      9238 - Social Service <br />
      9254 - Tourism <br />
      9500 - Nursing <br />
      9556 - Oral Hygiene <br />
      9670 - Advertising and Marketing Management <br />
      9773 - Journalism and Communication <br />
      9853 - Basic Education <br />
      9991 - Management (evening attendance)
      </td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Daytime/evening attendance</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 – daytime <br/> 0 - evening</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Previous qualification</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Education Level</td>
      <td>1 - Secondary education <br/>
      2 - Higher education - bachelor's degree <br/>
      3 - Higher education - degree <br/>
      4 - Higher education - master's <br/>
      5 - Higher education - doctorate <br/>
      6 - Frequency of higher education <br/>
      9 - 12th year of schooling - not completed <br/>
      10 - 11th year of schooling - not completed <br/>
      12 - Other - 11th year of schooling <br/>
      14 - 10th year of schooling <br/>
      15 - 10th year of schooling - not completed <br/>
      19 - Basic education 3rd cycle (9th/10th/11th year) or equiv. <br/>
      38 - Basic education 2nd cycle (6th/7th/8th year) or equiv. <br/>
      39 - Technological specialization course <br/>
      40 - Higher education - degree (1st cycle) <br/>
      42 - Professional higher technical course <br/>
      43 - Higher education - master (2nd cycle)
      </td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Previous qualification (grade)</td>
      <td>Feature</td>
      <td>Continuous</td>
      <td></td>
      <td>Grade of previous qualification (between 0 and 200)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Nacionality</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Nationality</td>
      <td>1 - Portuguese<br/>
      2 - German<br/>
      6 - Spanish<br/>
      11 - Italian<br/>
      13 - Dutch<br/>
      14 - English<br/>
      17 - Lithuanian<br/>
      21 - Angolan<br/>
      22 - Cape Verdean<br/>
      24 - Guinean<br/>
      25 - Mozambican<br/>
      26 - Santomean<br/>
      32 - Turkish<br/>
      41 - Brazilian<br/>
      62 - Romanian<br/>
      100 - Moldova (Republic of)<br/>
      101 - Mexican<br/>
      103 - Ukrainian<br/>
      105 - Russian<br/>
      108 - Cuban<br/>
      109 - Colombian
      </td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Mother's qualification</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Education Level</td>
      <td>1 - Secondary Education - 12th Year of Schooling or Eq. <br/>
      2 - Higher Education - Bachelor's Degree <br/>
      3 - Higher Education - Degree <br/>
      4 - Higher Education - Master's <br/>
      5 - Higher Education - Doctorate <br/>
      6 - Frequency of Higher Education <br/>
      9 - 12th Year of Schooling - Not Completed <br/>
      10 - 11th Year of Schooling - Not Completed <br/>
      11 - 7th Year (Old) <br/>
      12 - Other - 11th Year of Schooling <br/>
      14 - 10th Year of Schooling <br/>
      18 - General commerce course <br/>
      19 - Basic Education 3rd Cycle (9th/10th/11th Year) or Equiv. <br/>
      22 - Technical-professional course <br/>
      26 - 7th year of schooling <br/>
      27 - 2nd cycle of the general high school course <br/>
      29 - 9th Year of Schooling - Not Completed <br/>
      30 - 8th year of schooling <br/>
      34 - Unknown <br/>
      35 - Can't read or write <br/>
      36 - Can read without having a 4th year of schooling <br/>
      37 - Basic education 1st cycle (4th/5th year) or equiv. <br/>
      38 - Basic Education 2nd Cycle (6th/7th/8th Year) or Equiv. <br/>
      39 - Technological specialization course <br/>
      40 - Higher education - degree (1st cycle) <br/>
      41 - Specialized higher studies course <br/>
      42 - Professional higher technical course <br/>
      43 - Higher Education - Master (2nd cycle) <br/>
      44 - Higher Education - Doctorate (3rd cycle)
      </td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Father's qualification</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Education Level</td>
      <td>1 - Secondary Education - 12th Year of Schooling or Eq. <br/>
      2 - Higher Education - Bachelor's Degree <br/>
      3 - Higher Education - Degree <br/>
      4 - Higher Education - Master's <br/>
      5 - Higher Education - Doctorate <br/>
      6 - Frequency of Higher Education <br/>
      9 - 12th Year of Schooling - Not Completed <br/>
      10 - 11th Year of Schooling - Not Completed <br/>
      11 - 7th Year (Old) <br/>
      12 - Other - 11th Year of Schooling <br/>
      13 - 2nd year complementary high school course <br/>
      14 - 10th Year of Schooling <br/>
      18 - General commerce course <br/>
      19 - Basic Education 3rd Cycle (9th/10th/11th Year) or Equiv. <br/>
      20 - Complementary High School Course <br/>
      22 - Technical-professional course <br/>
      25 - Complementary High School Course - not concluded <br/>
      26 - 7th year of schooling <br/>
      27 - 2nd cycle of the general high school course <br/>
      29 - 9th Year of Schooling - Not Completed <br/>
      30 - 8th year of schooling <br/>
      31 - General Course of Administration and Commerce <br/>
      33 - Supplementary Accounting and Administration <br/>
      34 - Unknown <br/>
      35 - Can't read or write <br/>
      36 - Can read without having a 4th year of schooling <br/>
      37 - Basic education 1st cycle (4th/5th year) or equiv. <br/>
      38 - Basic Education 2nd Cycle (6th/7th/8th Year) or Equiv. <br/>
      39 - Technological specialization course <br/>
      40 - Higher education - degree (1st cycle) <br/>
      41 - Specialized higher studies course <br/>
      42 - Professional higher technical course <br/>
      43 - Higher Education - Master (2nd cycle) <br/>
      44 - Higher Education - Doctorate (3rd cycle)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Mother's occupation</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Occupation</td>
      <td>0 - Student <br/>
      1 - Representatives of the Legislative Power and Executive Bodies, Directors, Directors and Executive Managers <br/>
      2 - Specialists in Intellectual and Scientific Activities <br/>
      3 - Intermediate Level Technicians and Professions <br/>
      4 - Administrative staff <br/>
      5 - Personal Services, Security and Safety Workers and Sellers <br/>
      6 - Farmers and Skilled Workers in Agriculture, Fisheries and Forestry <br/>
      7 - Skilled Workers in Industry, Construction and Craftsmen <br/>
      8 - Installation and Machine Operators and Assembly Workers <br/>
      9 - Unskilled Workers <br/>
      10 - Armed Forces Professions <br/>
      90 - Other Situation <br/>
      99 - (blank) <br/>
      122 - Health professionals <br/>
      123 - teachers <br/>
      125 - Specialists in information and communication technologies (ICT) <br/>
      131 - Intermediate level science and engineering technicians and professions <br/>
      132 - Technicians and professionals, of intermediate level of health <br/>
      134 - Intermediate level technicians from legal, social, sports, cultural and similar services <br/>
      141 - Office workers, secretaries in general and data processing operators <br/>
      143 - Data, accounting, statistical, financial services and registry-related operators <br/>
      144 - Other administrative support staff <br/>
      151 - personal service workers <br/>
      152 - sellers <br/>
      153 - Personal care workers and the like <br/>
      171 - Skilled construction workers and the like, except electricians <br/>
      173 - Skilled workers in printing, precision instrument manufacturing, jewelers, artisans and the like <br/>
      175 - Workers in food processing, woodworking, clothing and other industries and crafts <br/>
      191 - cleaning workers <br/>
      192 - Unskilled workers in agriculture, animal production, fisheries and forestry <br/>
      193 - Unskilled workers in extractive industry, construction, manufacturing and transport <br/>
      194 - Meal preparation assistants</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Father's occupation</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Occupation</td>
      <td>0 - Student <br/>
      1 - Representatives of the Legislative Power and Executive Bodies, Directors, Directors and Executive Managers <br/>
      2 - Specialists in Intellectual and Scientific Activities <br/>
      3 - Intermediate Level Technicians and Professions <br/>
      4 - Administrative staff <br/>
      5 - Personal Services, Security and Safety Workers and Sellers <br/>
      6 - Farmers and Skilled Workers in Agriculture, Fisheries and Forestry <br/>
      7 - Skilled Workers in Industry, Construction and Craftsmen <br/>
      8 - Installation and Machine Operators and Assembly Workers <br/>
      9 - Unskilled Workers <br/>
      10 - Armed Forces Professions <br/>
      90 - Other Situation <br/>
      99 - (blank) <br/>
      101 - Armed Forces Officers <br/>
      102 - Armed Forces Sergeants <br/>
      103 - Other Armed Forces personnel <br/>
      112 - Directors of administrative and commercial services <br/>
      114 - Hotel, catering, trade and other services directors <br/>
      121 - Specialists in the physical sciences, mathematics, engineering and related techniques <br/>
      122 - Health professionals <br/>
      123 - teachers <br/>
      124 - Specialists in finance, accounting, administrative organization, public and commercial relations <br/>
      131 - Intermediate level science and engineering technicians and professions <br/>
      132 - Technicians and professionals, of intermediate level of health <br/>
      134 - Intermediate level technicians from legal, social, sports, cultural and similar services <br/>
      135 - Information and communication technology technicians <br/>
      141 - Office workers, secretaries in general and data processing operators <br/>
      143 - Data, accounting, statistical, financial services and registry-related operators <br/>
      144 - Other administrative support staff <br/>
      151 - personal service workers <br/>
      152 - sellers <br/>
      153 - Personal care workers and the like <br/>
      154 - Protection and security services personnel <br/>
      161 - Market-oriented farmers and skilled agricultural and animal production workers <br/>
      163 - Farmers, livestock keepers, fishermen, hunters and gatherers, subsistence <br/>
      171 - Skilled construction workers and the like, except electricians <br/>
      172 - Skilled workers in metallurgy, metalworking and similar <br/>
      174 - Skilled workers in electricity and electronics <br/>
      175 - Workers in food processing, woodworking, clothing and other industries and crafts <br/>
      181 - Fixed plant and machine operators <br/>
      182 - assembly workers <br/>
      183 - Vehicle drivers and mobile equipment operators <br/>
      192 - Unskilled workers in agriculture, animal production, fisheries and forestry <br/>
      193 - Unskilled workers in extractive industry, construction, manufacturing and transport <br/>
      194 - Meal preparation assistants <br/>
      195 - Street vendors (except food) and street service providers
      </td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Admission grade</td>
      <td>Feature</td>
      <td>Continuous</td>
      <td></td>
      <td>Admission grade (between 0 and 200)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Displaced</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 – yes <br/>
      0 – no</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Educational special needs</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 – yes <br/>
      0 – no</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Debtor</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 – yes <br/>
      0 – no</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Tuition fees up to date</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 – yes <br/>
      0 – no</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Gender</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Gender</td>
      <td>1 – male <br/>
      0 – female</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Scholarship holder</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 – yes <br/>
      0 – no</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Age at enrollment</td>
      <td>Feature</td>
      <td>Integer</td>
      <td>Age</td>
      <td>Age of studend at enrollment</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>International</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>1 – yes <br/>
      0 – no</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 1st sem (credited)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units credited in the 1st semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 1st sem (enrolled)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units enrolled in the 1st semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 1st sem (evaluations)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of evaluations to curricular units in the 1st semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 1st sem (approved)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units approved in the 1st semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 1st sem (grade)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Grade average in the 1st semester (between 0 and 20)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 1st sem (without evaluations)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units without evalutions in the 1st semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 2nd sem (credited)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units credited in the 2nd semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 2nd sem (enrolled)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units enrolled in the 2nd semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 2nd sem (evaluations)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of evaluations to curricular units in the 2nd semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 2nd sem (approved)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units approved in the 2nd semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 2nd sem (grade)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Grade average in the 2nd semester (between 0 and 20)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Curricular units 2nd sem (without evaluations)</td>
      <td>Feature</td>
      <td>Integer</td>
      <td></td>
      <td>Number of curricular units without evalutions in the 1st semester</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Unemployment rate</td>
      <td>Feature</td>
      <td>Continuous</td>
      <td></td>
      <td>Unemployment rate (%)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Inflation rate</td>
      <td>Feature</td>
      <td>Continuous</td>
      <td></td>
      <td>Inflation rate (%)</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>GDP</td>
      <td>Feature</td>
      <td>Continuous</td>
      <td></td>
      <td>GDP</td>
      <td></td>
      <td>no</td>
    </tr>
    <tr>
      <td>Target</td>
      <td>Target</td>
      <td>Categorical</td>
      <td></td>
      <td>Target. The problem is formulated as a three category classification task (dropout, enrolled, and graduate) at the end of the normal duration of the course</td>
      <td></td>
      <td>no</td>
    </tr>
  </tbody>
</table>

According to the table above, there are no missing values in any of the columns. Datatypes are also given in the table. As we do not use the original data but use the version generated by Kaggle, we validate that datatypes are actually the same and there are no missing values, using the function `info()`. Indeed, datatypes are `int` and `float` and there are no null values as there are 76518 entries and for each column the number of non-null values is also 76518.

In [ ]:
train.info()

Outliers detection with subsequent imputation plays an important role in training a machine learning model. If outliers are not treated properly, the quality of a trained model may be very low. For a starting point, it's worth using the `describe()` function in order to get an overview of a training set. Usually it is possible to see whether the values make any sense or not (e.g. negative distance between two cities), paying attention to statistics such as `max` or `min`.

In [ ]:
train.describe()

No outliers were detected in our analysis. However, several points are worth noting:
1. The variable description table shows that many columns represent categorical codes, even though they use integer values. This makes outlier detection challenging, especially in the context of our project. For such columns, which are more categorical than numerical, the only feasible validation is to confirm that all values align with those defined in the variable descriptions. Outlier detection through unrealistic combinations of multiple features could be explored, but in our case, there are no clear contradictions between features (e.g., marital status is independent of nationality; a mother’s occupation is unrelated to a student's gender, etc.).
2. Some columns in the training set, specifically numerical ones like `age at enrollment`, `unemployment rate`, `inflation rate`, `GDP`, `admission grade`, and `previous qualification (grade)`, could potentially contain outliers. We focused particularly on these features but found no outliers in this training set.
3. A "rigorous data preprocessing" has already been conducted on the original data (from which the training set was generated) by the authors of the initial study <a url="https://archive.ics.uci.edu/dataset/697/predict+students+dropout+and+academic+success">[1]</a>. This is most likely the reason why the given training set does not contain any outliers either.

Two issues need to be addressed:
1. The `id` column does not contain meaningful information for the model, so it can be removed.
2. According to the variable description table, many columns represent categorical codes. Converting these values to strings is advisable so that ML algorithms do not treat them as numerical values with potential significance, avoiding using numerical semantics that lack true meaning.

In [6]:
# Drop column 'id' in both train and test
train.drop("id", axis=1, inplace=True)
test.drop("id", axis=1, inplace=True)

In [7]:
# Cast columns with categorical codes to strings
numerical_features = [
    "Previous qualification (grade)",
    "Admission grade",
    "Age at enrollment",
    "Curricular units 1st sem (credited)",
    "Curricular units 1st sem (enrolled)",
    "Curricular units 1st sem (evaluations)",
    "Curricular units 1st sem (approved)",
    "Curricular units 1st sem (grade)",
    "Curricular units 1st sem (without evaluations)",
    "Curricular units 2nd sem (credited)",
    "Curricular units 2nd sem (enrolled)",
    "Curricular units 2nd sem (evaluations)",
    "Curricular units 2nd sem (approved)",
    "Curricular units 2nd sem (grade)",
    "Curricular units 2nd sem (without evaluations)",
    "Unemployment rate",
    "Inflation rate",
    "GDP"
]

categorical_features = list(train.columns)
for col in numerical_features:
    categorical_features.remove(col)
categorical_features.remove("Target")

assert len(categorical_features) + len(numerical_features) + 1 == len(train.columns)

for col in categorical_features:
    train[col] = train[col].astype(str)
    test[col] = test[col].astype(str)

train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76518 entries, 0 to 76517
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  76518 non-null  object 
 1   Application mode                                76518 non-null  object 
 2   Application order                               76518 non-null  object 
 3   Course                                          76518 non-null  object 
 4   Daytime/evening attendance                      76518 non-null  object 
 5   Previous qualification                          76518 non-null  object 
 6   Previous qualification (grade)                  76518 non-null  float64
 7   Nacionality                                     76518 non-null  object 
 8   Mother's qualification                          76518 non-null  object 
 9   Father's qualification                 

<a name="exploratory-data-analysis"></a>
## Exploratory data analysis

In [ ]:
# Imports for visualization
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns

Outliers detection with subsequent imputation plays an important role in training a machine learning model. If outliers are not treated properly, the quality of a trained model may be very low. For a starting point, it's worth using the `describe()` function in order to get an overview of a training set. Usually it is possible to see whether the values make any sense or not (e.g. negative distance between two cities), paying attention to statistics such as `max` or `min`.

In [ ]:
train.describe()

No outliers were detected in our analysis. However, several points are worth noting:
1. The variable description table shows that many columns represent categorical codes, even though they use integer values. This makes outlier detection challenging, especially in the context of our project. For such columns, which are more categorical than numerical, the only feasible validation is to confirm that all values align with those defined in the variable descriptions. Outlier detection through unrealistic combinations of multiple features could be explored, but in our case, there are no clear contradictions between features (e.g., marital status is independent of nationality; a mother’s occupation is unrelated to a student's gender, etc.).
2. Some columns in the training set, specifically numerical ones like `age at enrollment`, `unemployment rate`, `inflation rate`, `GDP`, `admission grade`, and `previous qualification (grade)`, could potentially contain outliers. We focused particularly on these features but found no outliers in this training set.
3. A "rigorous data preprocessing" has already been conducted on the original data (from which the training set was generated) by the authors of the initial study <a url="https://archive.ics.uci.edu/dataset/697/predict+students+dropout+and+academic+success">[1]</a>. This is most likely the reason why the given training set does not contain any outliers either.

Next, we aim to visualize the relationships between specific features and target-class. For this, we chose such features as `Gender`, `Marital Status`, `Admission Grade`, `Age at Enrollment`, and some others. We hypothesize that these features might be the strongest predictors of the `Target`, though we acknowledge the possibility of bias in our assumption.

In [ ]:
ax = sns.countplot(x="Target", hue="Gender", data=train)
plt.legend(labels = ['female', 'male'])
sns.move_legend(ax, "upper left", bbox_to_anchor=(1, 1))
plt.show()

In [ ]:
ax = sns.countplot(x="Target", hue="Marital status", data=train)
#plt.legend(labels = ['female', 'male'])
sns.move_legend(ax, "upper left", bbox_to_anchor=(1, 1))
plt.show()

In [ ]:
idx = 0
colors = ['red', 'gold', 'green']

for target, group in train.groupby('Target'):
    sns.kdeplot(group[["Admission grade"]].squeeze(), label=f'{target}', color=colors[idx])
    idx += 1

plt.legend()
plt.show()

In [ ]:
idx = 0
colors = ['red', 'gold', 'green']

for target, group in train.groupby('Target'):
    sns.kdeplot(group[["Age at enrollment"]].squeeze(), label=f'{target}', color=colors[idx])
    idx += 1

plt.legend()
plt.show()

In [ ]:
idx = 0
colors = ["red", "gold", "green"]

for target, group in train.groupby("Target"):
    sns.kdeplot(group[["Curricular units 1st sem (grade)"]].squeeze(), label=f"{target}", color=colors[idx])
    idx += 1

plt.legend()
plt.show()

In [ ]:
idx = 0
colors = ["red", "gold", "green"]

for target, group in train.groupby("Target"):
    sns.kdeplot(group[["Curricular units 2nd sem (grade)"]].squeeze(), label=f"{target}", color=colors[idx])
    idx += 1

plt.legend()
plt.show()

For a high-quality ML model, it is crucial to verify the balance of the training set. If the training data is imbalanced, the model may fail to adequately learn from or attend to underrepresented target classes, leading to biased predictions and reduced performance.

In [ ]:
train["Target"].value_counts().plot(kind="bar")

In [ ]:
train["Target"].value_counts()

Target
Graduate    36282
Dropout     25296
Enrolled    14940
Name: count, dtype: int64

The output above shows that there are significantly more rows with the target "Graduate" compared to "Dropout" and "Enrolled." Thus, "Dropout" and "Enrolled" are underrepresented, and we need to apply an oversampling technique. In the following cell, we use SMOTE to address this imbalance. The output confirms that the classes are now balanced.

In [8]:
from imblearn.over_sampling import SMOTENC

smote = SMOTENC(random_state=42, categorical_features=categorical_features)
X_train, y_train = smote.fit_resample(train.drop("Target", axis=1), train["Target"])
train = pd.concat([X_train, y_train], axis=1)
train["Target"].value_counts()

Target
Graduate    36282
Dropout     36282
Enrolled    36282
Name: count, dtype: int64

In [9]:
X_test = test # Rename it to be more stylish

#### Visualizing distribution of values by target class for all columns

In [ ]:
# Define the color palette and the fixed order for the target categories
target_order = ["Dropout", "Enrolled", "Graduate"]
colors = ["red", "gold", "green"]

# Define line styles for the target categories
line_styles = ["solid", "dashed", "dotted"]

In [ ]:
### Plots for Numerical Features

max_plots_per_row = 3
num_cols = len(numerical_features)
num_rows = (num_cols + max_plots_per_row - 1) // max_plots_per_row  # Ensures enough rows for all plots

fig, axes = plt.subplots(num_rows, max_plots_per_row, figsize=(20, 6 * num_rows))
axes = axes.flatten()

# General title
fig.suptitle("Density Distribution of Numerical Values by Target Class", fontsize=20, y=1.0)


for i, col in enumerate(numerical_features):
    bins = train[col].max() - train[col].min() + 1

    # Calculate bin edges if column is integer type
    if pd.api.types.is_integer_dtype(train[col]):
        min_value = train[col].min()
        max_value = train[col].max()

        # Define bins so that each integer value will have their own bin
        bin_edges = np.linspace(min_value, max_value, max_value - min_value + 1)

    for idx, target in enumerate(target_order):
        # Filter the group based on the defined target order
        group = train[train["Target"] == target]

        # Greate graph based on numberic type
        if pd.api.types.is_integer_dtype(train[col]):
            sns.histplot(data=group, x=col, label=target, color=colors[idx], ax=axes[i],
                     kde=False, stat="density", bins=bin_edges, element="step", fill=False, linestyle=line_styles[idx % len(line_styles)])
        elif pd.api.types.is_float_dtype(train[col]):
            sns.kdeplot(group[col].squeeze(), label=target, color=colors[idx], ax=axes[i], linestyle=line_styles[idx % len(line_styles)])
        else:
            raise Exception(f"{col} is not a numerical type")


    # Set titles and labels
    axes[i].set_title(f"Distribution of '{col}' by Target Class")
    axes[i].set_xlabel(col)
    axes[i].set_ylabel("Density")
    axes[i].legend(title="Target", loc="upper right")

# Hide any unused subplots if num_cols is not a perfect multiple of max_plots_per_row
for j in range(i + 1, len(axes)):
    fig.delaxes(axes[j])

plt.tight_layout()
plt.show()

In [ ]:
### Plots for categorical features

max_plots_per_row = 3
num_cols = len(categorical_features)
num_rows = (num_cols // max_plots_per_row) + (1 if num_cols % max_plots_per_row != 0 else 0)

fig, axes = plt.subplots(num_rows, max_plots_per_row, figsize=(20, 6 * num_rows))
axes = axes.flatten()

# General title
fig.suptitle("Density Distribution of Categorical Values by Target Class", fontsize=20, y=1.0)

for i, col in enumerate(categorical_features):
    # Get the value counts and calculate cumulative frequency
    value_counts = train[col].value_counts(normalize=True)
    cumulative_frequency = value_counts.cumsum()

    # Determine total unique values
    total_unique_values = train[col].nunique()

    # Show only top max 10 values
    num_values_to_show = min(10, total_unique_values)

    # Get the top values based on the calculated number to display
    top_values = value_counts.nlargest(num_values_to_show).index.tolist()
    train_top = train[train[col].isin(top_values)]

    # Calculate counts for each category within each target class
    counts = train_top.groupby(["Target", col]).size().reset_index(name="Count")

    # Normalize counts to get proportions within each target group
    total_counts = counts.groupby("Target")["Count"].transform("sum")
    counts["Proportion"] = counts["Count"] / total_counts

    # Plot proportions
    sns.barplot(
        data=counts, x=col, y="Proportion", hue="Target", ax=axes[i],
        palette=colors, hue_order=target_order, order=top_values
    )

    # Set titles and labels
    axes[i].set_title(f"Distribution of '{col}' by Target Class (top {num_values_to_show}/{total_unique_values})")
    axes[i].set_xlabel(col)
    axes[i].set_ylabel("Density")
    axes[i].legend(title="Target", loc="upper right")

# Hide any unused subplots if num_cols is not a perfect multiple of max_plots_per_row
for j in range(i + 1, len(axes)):
    fig.delaxes(axes[j])

plt.tight_layout()
plt.show()


<a name="feature-engineering-and-selection"></a>
## Feature engineering and selection

<a name="baseline-model-establishing"></a>
## Baseline model establishing

We established two baseline models. The first model predicts the target classes entirely at random, using the function `random_predict` (see below). With this function, we achieved a private score (accuracy) of 0.33501 and a public score of 0.33542.

In [ ]:
test_first_id = 76518

def random_predict(test):
  np.random.seed(42)
  return np.random.choice(["Graduate", "Dropout", "Enrolled"], len(test))

random_predict_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                          'Target': random_predict(test)})
random_predict_submission.to_csv('random_predict_submission.csv', index=False)

!kaggle competitions submit -c playground-series-s4e6 -f random_predict_submission.csv -m "Submission using random_predict"

We also tested another baseline model that predicts the target classes randomly, but with the same distribution as the training data, accounting for the observed class imbalance. This model is implemented as the function `random_predict_with_dist` (see below). Using this function, we achieved a private score of 0.37522 and a public score of 0.36865.

In [ ]:
def random_predict_with_dist(test):
  np.random.seed(42)
  classes = ["Graduate", "Dropout", "Enrolled"]
  num_per_class = [36282, 25296, 14940]
  num_total = sum(num_per_class)
  distribution = [num / num_total for num in num_per_class]
  return np.random.choice(classes, len(test), p=distribution)

random_predict_with_dist_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                                    'Target': random_predict_with_dist(test)})
random_predict_with_dist_submission.to_csv('random_predict_with_dist_submission.csv', index=False)

!kaggle competitions submit -c playground-series-s4e6 -f random_predict_with_dist_submission.csv -m "Submission using random_predict_with_dist"

A more meaningful baseline model might be a standard Logistic Regression. Unlike random guessing, this model learns from the training data to make predictions. With this model, we achieved a private score of 0.76505 and a public score of 0.76220. While these scores may seem reasonable, they are actually quite low: in a competition, this performance would rank around 2497th out of 2685. For comparison, the first-place model has an accuracy of 0.84035.

In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(train.drop("Target", axis=1), train["Target"])

logistic_regression_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                                'Target': model.predict(test)})
logistic_regression_submission.to_csv('logistic_regression_submission.csv', index=False)

!kaggle competitions submit -c playground-series-s4e6 -f logistic_regression_submission.csv -m "Submission using logistic regression"

In the following section, we will explore more sophisticated models with the goal of achieving significantly better performance than the baseline models.

<a name="models-comparison"></a>
## Models comparison

We aim to get improved results. To start, we try stacking using two models: Logistic Regression and Random Forest Classifier. For final estimator we use Support Vector Machine, namely Linear Support Vector Classification. In the next cell we fit the model.

In [ ]:
from sklearn.ensemble import StackingClassifier
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

estimators = [
    ('log_reg', LogisticRegression()),
    ('random_forest', RandomForestClassifier())
]

np.random.seed(42)
stacking_classifier = StackingClassifier(estimators=estimators, final_estimator=LinearSVC(), cv=4)
stacking_classifier.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

KeyboardInterrupt: 

In the next cell we predict classes for test dataset and submit it to Kaggle.

In [ ]:
y_pred = stacking_classifier.predict(X_test)
test_first_id = 76518
stacking_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                    'Target': y_pred})

stacking_submission.to_csv('stacking_submission.csv', index=False)
!kaggle competitions submit -c playground-series-s4e6 -f stacking_submission.csv -m "Submission using stacking"

With this model, we achieved a private score of 0.82271 and a public score of 0.82474, which would place us at rank 2193. While this meets the minimum goal (according to our plan the minimum goal was at least 0.80000), there's still significant work ahead to reach the ultimate target (accuracy 0.83500 or higher).

<img src="https://i.imgur.com/lgH4FdB.png"/>

We also tried GradientBoostingClassifier. Unfortunately, this gave even worse accuracy of 0.82032.

<img src="https://i.imgur.com/cvKNamj.png" />

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

gb_model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)
gb_model.fit(X_train, y_train)

KeyboardInterrupt: 

In [ ]:
y_pred = gb_model.predict(X_test)
test_first_id = 76518
gb_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                    'Target': y_pred})

gb_submission.to_csv('gb_submission.csv', index=False)
!kaggle competitions submit -c playground-series-s4e6 -f gb_submission.csv -m "Submission using GB"

100% 760k/760k [00:00<00:00, 2.00MB/s]
Successfully submitted to Classification with an Academic Success Dataset

And we also tried AdaBoostClassifier. This game us accuracy of 0.81238, which is not an improvement :(

<img src="https://i.imgur.com/X9MJ6TU.png" />

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

ab_model = AdaBoostClassifier(n_estimators=50, learning_rate=1.0, random_state=42)
ab_model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


AdaBoostClassifier(random_state=42)

In [ ]:
y_pred = ab_model.predict(X_test)
test_first_id = 76518
ab_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                    'Target': y_pred})

ab_submission.to_csv('ab_submission.csv', index=False)
!kaggle competitions submit -c playground-series-s4e6 -f ab_submission.csv -m "Submission using AB"

100% 760k/760k [00:00<00:00, 1.87MB/s]
Successfully submitted to Classification with an Academic Success Dataset

Of course, participating in a Kaggle competition means you should try XGBoost. We had high hopes for this model. But it gave almost the worst accuracy, being better than LogisticRegression only by ~1%. Namely, using XGB gave us public score of 0.79523.

<img src="https://i.imgur.com/p4qWlV7.png"/>

TLDR:

<img src="https://www.epmguidance.com/wp-content/uploads/2010/03/no-free-lunch.png" width="300"/>

In [ ]:
import xgboost as xgb

target_mapping = {'Dropout': 0, 'Enrolled': 1, 'Graduate': 2}
y_train_xgb = train['Target'].map(target_mapping)
X_train_xgb = train.drop('Target', axis=1)
for col in X_train_xgb.select_dtypes(include=['object']).columns:
    X_train_xgb[col] = X_train_xgb[col].astype('category')


xgb_model = xgb.XGBClassifier(n_estimators=100, learning_rate=0.01, random_state=42, enable_categorical=True, objective="multi:softmax")
xgb_model.fit(X_train_xgb, y_train_xgb)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=True, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.01, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=100, n_jobs=None,
              num_parallel_tree=None, objective='multi:softmax', ...)

In [ ]:
X_test_xgb = X_test.copy()
for col in X_test_xgb.select_dtypes(include=['object']).columns:
    X_test_xgb[col] = X_test_xgb[col].astype('category')

y_pred = xgb_model.predict(X_test_xgb)

target_mapping_reverse = {v: k for k, v in target_mapping.items()}
y_pred = np.array([target_mapping_reverse[pred] for pred in y_pred])

test_first_id = 76518
xgb_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                    'Target': y_pred})

xgb_submission.to_csv('xgb_submission.csv', index=False)
!kaggle competitions submit -c playground-series-s4e6 -f xgb_submission.csv -m "Submission using XGB"

100% 760k/760k [00:00<00:00, 2.11MB/s]
Successfully submitted to Classification with an Academic Success Dataset

We had hopes that an ensemble of GradientBoostingClassifier, AdaBoostClassifier and XGBClassified could do some better work. However, an ensemble of these models was worse than GradientBoostingClassifier solely. Accuracy was 0.81954.

<img src="https://i.imgur.com/qubkruB.png" />

In [10]:
estimators = [
    ("gb", GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)),
    ("ab", AdaBoostClassifier(n_estimators=50, learning_rate=1.0, random_state=42)),
    ("xgb", xgb.XGBClassifier(n_estimators=100, learning_rate=0.01, random_state=42, enable_categorical=True, objective="multi:softmax"))
]

np.random.seed(42)
ensemble_classifier = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression(), cv=4)
ensemble_classifier.fit(X_train_xgb, y_train_xgb)

NameError: name 'GradientBoostingClassifier' is not defined

In [ ]:
y_pred = ensemble_classifier.predict(X_test_xgb)
y_pred = np.array([target_mapping_reverse[pred] for pred in y_pred])

test_first_id = 76518
ensemble_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                    'Target': y_pred})

ensemble_submission.to_csv('ensemble_submission.csv', index=False)
!kaggle competitions submit -c playground-series-s4e6 -f ensemble_submission.csv -m "Submission using Ensemble"

100% 760k/760k [00:00<00:00, 1.87MB/s]
Successfully submitted to Classification with an Academic Success Dataset

Finally, we decided to try out something fancy. Namely, we tried out such tool as AutoGluon (https://auto.gluon.ai/stable/index.html).
> AutoGluon is an open-source machine learning library developed by Amazon that focuses on automating machine learning (AutoML) tasks. It is designed to simplify the process of building, training, and deploying machine learning models.

Main features of it are:
* Quick Prototyping - Build machine learning solutions on raw data in a few lines of code.
* State-of-the-art Techniques - Automatically utilize SOTA models without expert knowledge.
* Easy to Deploy - Move from experimentation to production with cloud predictors and pre-built containers.
* Customizable - Extensible with custom feature processing, models, and metrics.


Training a model with Gluon took more time but eventually it gave some good results. We used Gluon with default settings (actually, we tried to use some of its parameters such as `presets`, `eval_metric` and others but this didn't improve the results) and this tool came up with a model that received accuracy of 0.83336.

<img src="https://i.imgur.com/jQwFtMs.png" />

In [ ]:
!pip install autogluon

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 1.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 1.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.0/117.0 kB 4.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 259.5/259.5 kB 8.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
INFO: pip is looking at multiple versions of openxlab to determine which version is compatible with other requirements. This could take a while.
INFO: pip is still looking at multiple versions of openxlab to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort 

In [ ]:
from autogluon.tabular import TabularDataset, TabularPredictor

train_data = TabularDataset(train)
test_data = TabularDataset(test)

predictor = TabularPredictor(label='Target').fit(train_data=train_data)
predictions = predictor.predict(test_data)

No path specified. Models will be saved in: "AutogluonModels/ag-20241215_130324"
Verbosity: 2 (Standard Logging)
=================== System Info ===================
AutoGluon Version:  1.2
Python Version:     3.10.12
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP PREEMPT_DYNAMIC Thu Jun 27 21:05:47 UTC 2024
CPU Count:          2
Memory Avail:       9.35 GB / 12.67 GB (73.8%)
Disk Space Avail:   68.75 GB / 107.72 GB (63.8%)
No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets. Defaulting to `'medium'`...
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='experimental' : New in v1.2: Pre-trained foundation model + parallel fits. The absolute best accuracy without consideration for inference speed. Does not support GPU.
	presets='best'         : Maximize accuracy. Recommended for most users. Use in competitions a

[1000]	valid_set's multi_logloss: 0.402882	valid_set's f1_weighted: 0.838386
[2000]	valid_set's multi_logloss: 0.392523	valid_set's f1_weighted: 0.843155
[3000]	valid_set's multi_logloss: 0.388143	valid_set's f1_weighted: 0.850008
[4000]	valid_set's multi_logloss: 0.386801	valid_set's f1_weighted: 0.853565
[5000]	valid_set's multi_logloss: 0.388362	valid_set's f1_weighted: 0.854055


	0.8552	 = Validation score   (f1_weighted)
	316.54s	 = Training   runtime
	10.97s	 = Validation runtime
Fitting model: LightGBM ...
	0.8441	 = Validation score   (f1_weighted)
	44.91s	 = Training   runtime
	0.54s	 = Validation runtime
Fitting model: RandomForestGini ...
	0.8366	 = Validation score   (f1_weighted)
	103.91s	 = Training   runtime
	0.37s	 = Validation runtime
Fitting model: RandomForestEntr ...
	0.8364	 = Validation score   (f1_weighted)
	121.08s	 = Training   runtime
	0.43s	 = Validation runtime
Fitting model: CatBoost ...


In [ ]:
test_first_id = 76518
gluon_submission = pd.DataFrame({'id': list(range(test_first_id, test_first_id + len(test))),
                                    'Target': predictions})

gluon_submission.to_csv('gluon_submission.csv', index=False)
!kaggle competitions submit -c playground-series-s4e6 -f gluon_submission.csv -m "Submission using Gluon"

100% 758k/758k [00:00<00:00, 1.65MB/s]
Successfully submitted to Classification with an Academic Success Dataset

<a name="hyperparameter-tuning"></a>
## Hyperparameter tuning

<a name="model-evaluation"></a>
## Model evaluation

<a name="results-interpretation"></a>
## Results interpretation

<a name="conclusion"></a>
## Conclusion